### ISSUE TITLE
How to make flapMap lazy?

### ISSUE THREAD

**Issue by yuyanjulia** [nus-cs2030/2122-s1#175]
I have completed the lab by making Logger and map() lazy. But I think my flatMap still isn't lazy enough due to the wildcard.
The return type of flatMap is `? extends Logger<? extends U>>` right? When I tried to first `flatMapper.apply(this.value.get())` then finally return a new Logger, I received a lot of compile errors.
For example `Supplier<CAP#1> cannot be converted to Supplier<U>` and `where CAP#1,CAP#2 are fresh type-variables: ...`

So I wasn't sure how to use the wildcard properly. In the end I get the value from supplier first `U newValue = 
newLogger.getValue().get();` then construct a new Logger `new Logger<U>(() -> newValue, newLog);`
However, by doing this, my flapMap isn't lazy anymore. I am not sure if I have described my issue clearly.
Any help will be appreciated.

**Comment by optionalemon** [comment]
Hi julia,
I believe that return type of flatmap is `Logger<U>`, maybe you can try using that!
I have gotten the `type CAP#1` error as well, due to forgetting that flatmap actually maps to a logger than an actual value, not sure if that is what you are encoutering though :(

**Comment by yuyanjulia** [comment]
> Hi julia,
> I believe that return type of flatmap is `Logger<U>`, maybe you can try using that!
> I have gotten the `type CAP#1` error as well, due to forgetting that flatmap actually maps to a logger than an actual value, not sure if that is what you are encoutering though :(

Sorry I meant the output of flatMapper function is `? extends Logger<? extends U>>`? Then I want to return `Logger<U>` in the end. But I faced some difficulties to convert the result of flatMapper to `Logger<U>`

**Comment by c0j0s** [comment]
you can simply supply the entire flatmapper to your constructor to make it lazy, also flatmapper returns a subtype of logger, u will need to call a method in your logger class that returns the actual value e.g `get()`.
something like this `new Logger<U>(() -> flatmapper.apply(oldvalue).get(), newLog);`
since everything is in your supplier, it should be evaluated lazily.

**Comment by yuyanjulia** [comment]
> you can simply supply the entire flatmapper to your constructor to make it lazy, also flatmapper returns a subtype of logger, u will need to call a method in your logger class that returns the actual value e.g `get()`.
> something like this `new Logger<U>(() -> flatmapper.apply(oldvalue).get(), newLog);`
> since everything is in your supplier, it should be evaluated lazily.

I see! So it should work if I just call it in one line, instead of assigning it to `Supplier<U> newValue` then return `new Logger<U>(newValue, newLog)`

**Comment by c0j0s** [comment]
you can assign it to a separate `Supplier`, just make sure the faltmapper is within the supplier will do. otherwise it will trigger it eagerly.
like this `Supplier<U> newValue = () -> flatmapper.apply(oldvalue).get();`

**Comment by yuyanjulia** [comment]
> you can assign it to a separate `Supplier`, just make sure the faltmapper is within the supplier will do. otherwise it will trigger it eagerly.
> like this `Supplier<U> newValue = () -> flatmapper.apply(oldvalue).get();`

This didn't work for me as the compiler will tell me `Supplier<CAP#1> cannot be converted to Supplier<U>` ;-;

**Comment by c0j0s** [comment]
How did you define your flatmap function?
is it something like this?
`<U> Logger<U> flatMap(Function<? super T, ? extends Logger<? extends U>> mapper)`

**Comment by optionalemon** [comment]
> > Hi julia,
> > I believe that return type of flatmap is `Logger<U>`, maybe you can try using that!
> > I have gotten the `type CAP#1` error as well, due to forgetting that flatmap actually maps to a logger than an actual value, not sure if that is what you are encoutering though :(
> 
> Sorry I meant the output of flatMapper function is `? extends Logger<? extends U>>`? Then I want to return `Logger<U>` in the end. But I faced some difficulties to convert the result of flatMapper to `Logger<U>`

Hmmm I didn't use any`? extends Logger<? extends U>>` in any parts. Maybe its your wildcard that has been causing this error?

**Comment by yuyanjulia** [comment]
> How did you define your flatmap function?
> is it something like this?
> `<U> Logger<U> flatMap(Function<? super T, ? extends Logger<? extends U>> mapper)`

Yes! I just followed the instruction on codecrunch
And I did something like
`Logger<U> newLogger = flatMapper.apply(this.value.get());`
`Supplier<U> newValue = newLogger.getValue();`
This has some problem because `Logger<? extends U>` is not a subset of `Logger<U>`? 

But you are calling .get() too. Does it make flatmap eager?

**Comment by c0j0s** [comment]
> > How did you define your flatmap function?
> > is it something like this?
> > `<U> Logger<U> flatMap(Function<? super T, ? extends Logger<? extends U>> mapper)`
> 
> Yes! I just followed the instruction on codecrunch And I did something like `Logger<U> newLogger = flatMapper.apply(this.value.get());` `Supplier<U> newValue = newLogger.getValue();` This has some problem because `Logger<? extends U>` is not a subset of `Logger<U>`?
> 
> But you are calling .get() too. Does it make flatmap eager?

nope, the difference is that your flatMapper is outside of a supplier, it will be evaluated once it runs that line.

`Supplier<U> newValue = () -> flatmapper.apply(oldvalue).get();` 
this however wraps the mapper and get within a supplier (it's like supplying an instruction on what to run in the future, not actually running the code), it wouldn't trigger the get() inside if you didn't call supplier.get().

Logger<? extends U> should accept a logger of type U and subsets I believed.

**Comment by yuyanjulia** [comment]
> > > How did you define your flatmap function?
> > > is it something like this?
> > > `<U> Logger<U> flatMap(Function<? super T, ? extends Logger<? extends U>> mapper)`
> > 
> > 
> > Yes! I just followed the instruction on codecrunch And I did something like `Logger<U> newLogger = flatMapper.apply(this.value.get());` `Supplier<U> newValue = newLogger.getValue();` This has some problem because `Logger<? extends U>` is not a subset of `Logger<U>`?
> > But you are calling .get() too. Does it make flatmap eager?
> 
> nope, the difference is that your flatMapper is outside of a supplier, it will be evaluated once it runs that line.
> 
> `Supplier<U> newValue = () -> flatmapper.apply(oldvalue).get();` this however wraps the mapper and get within a supplier (it's like supplying an instruction on what to run in the future, not actually running the code), it wouldn't trigger the get() inside if you didn't call supplier.get().
> 
> Logger<? extends U> should accept a logger of type U and subsets I believed.

I see! So as long as the .get() is within the Supplier, it should still be lazy?

I also tried `Logger<? extends U> newLogger = flatMapper.apply(this.value.get());`
`Supplier<? extends U> newValue = newLogger.getValue();`
but when I try to return `new Logger<U>(newValue, newLog);` it has compile error again.

So I should still get() the value out first then assign it back into a supplier right?

**Comment by c0j0s** [comment]
> I see! So as long as the .get() is within the Supplier, it should still be lazy?
> 
> I also tried `Logger<? extends U> newLogger = flatMapper.apply(this.value.get());` `Supplier<? extends U> newValue = newLogger.getValue();` but when I try to return `new Logger<U>(newValue, newLog);` it has compile error again.
> 
> So I should still get() the value out first then assign it back into a supplier right?

nope, all the methods you called to retrieve the value should be wrapped inside the supplier. I'm not sure how you define your `getValue()`, for me, my get() return the actual value, not a supplier, but since my get is inside the supplier, it wouldn't evaluate it eagerly.

![image](https://user-images.githubusercontent.com/7760148/140309965-458355e2-bea5-49e4-9a45-a0fc5dc7925a.png)

**Comment by haoyusimon** [comment]
> > I see! So as long as the .get() is within the Supplier, it should still be lazy?
> > I also tried `Logger<? extends U> newLogger = flatMapper.apply(this.value.get());` `Supplier<? extends U> newValue = newLogger.getValue();` but when I try to return `new Logger<U>(newValue, newLog);` it has compile error again.
> > So I should still get() the value out first then assign it back into a supplier right?
> 
> nope, all the methods you called to retrieve the value should be wrapped inside the supplier. I'm not sure how you define your `getValue()`, for me, my get() return the actual value, not a supplier, but since my get is inside the supplier, it wouldn't evaluate it eagerly.
> 
> ![image](https://user-images.githubusercontent.com/7760148/140309965-458355e2-bea5-49e4-9a45-a0fc5dc7925a.png)

Sorry just to clarify, in your flatMap, is the mapper.apply() used twice? Then how did you manage to pass the last test case? I also used a similar approach but stuck at the last test case because I also used mapper.apply() twice.

**Comment by ff1Zzd** [comment]
> > > I see! So as long as the .get() is within the Supplier, it should still be lazy?
> > > I also tried `Logger<? extends U> newLogger = flatMapper.apply(this.value.get());` `Supplier<? extends U> newValue = newLogger.getValue();` but when I try to return `new Logger<U>(newValue, newLog);` it has compile error again.
> > > So I should still get() the value out first then assign it back into a supplier right?
> > 
> > 
> > nope, all the methods you called to retrieve the value should be wrapped inside the supplier. I'm not sure how you define your `getValue()`, for me, my get() return the actual value, not a supplier, but since my get is inside the supplier, it wouldn't evaluate it eagerly.
> > ![image](https://user-images.githubusercontent.com/7760148/140309965-458355e2-bea5-49e4-9a45-a0fc5dc7925a.png)
> 
> Hi c0j0s!! Sorry just to clarify, in your flatMap, is the mapper.apply() used twice? Then how did you manage to pass the last test case? I also used a similar approach but stuck at the last test case because I also used mapper.apply() twice.

Hi I am not sure whether this is the case but probably c0j0s has a different implementation for map. Since the last level only test for map anyway. For your problem, perhaps you can store the output of `mapper.apply(get())` into a single object, and then update your logHistory and value from that object. In this case, you will only need to call mapper.apply() once .

**Comment by c0j0s** [comment]
> Hi c0j0s!! Sorry just to clarify, in your flatMap, is the mapper.apply() used twice? Then how did you manage to pass the last test case? I also used a similar approach but stuck at the last test case because I also used mapper.apply() twice.

yup, I had the same issue as well, even with caching. basically, there's two separate get method called inside my toString() function, one for the logger[value] and the other for the history. what I did was to add a check in toString to see if there's value in my history, else get from my supplier. it's a workaround mentioned by others in another post I believed.

**Comment by throwawayAcc123** [comment]
Hi guys!! Just to clarify, is it possible to store all the log history in an ArrayList<String>, and only modify map() just to handle level 5 test case? Coz that's what I'm doing right now, but not sure if this idea of modifying just map() would work (coz I saw some discussion saying that the DS that stores log history has to be wrapped around Supplier 

any help is greatly aprpeciated!!

**Comment by c0j0s** [comment]
> Hi guys!! Just to clarify, is it possible to store all the log history in an ArrayList, and only modify map() just to handle level 5 test case? Coz that's what I'm doing right now, but not sure if this idea of modifying just map() would work (coz I saw some discussion saying that the DS that stores log history has to be wrapped around Supplier
> 
> any help is greatly aprpeciated!!

I don't think it's possible. wouldn't it affect the map() for previous levels? and without suppliers how do you avoid the double printing in level 5 if you call `mapper.apply()` eagerly, cos you surely will need the value in both `Logger[value]` and log history right?

**Comment by chenghong99** [comment]
What I did was to create an anonymous inner class that is wrapped in a supplier. Inside this wrapped class I performed al sorts of operation needed for flat map. The style of programming is similar for flatmap and map. I think it is also not a requirement for flatmap to be lazy though since there is no test case that ensures flatmap must be lazy. It would not be too tough to make flatmap lazy if you have made map lazy, since the idea behind both of them is the same.

**Comment by callmeqimin** [comment]
i think the issue appeared as you instantiate a class with wild class, you should :
![image](https://user-images.githubusercontent.com/26117678/140682436-02c4f906-9072-4143-ab6f-9fffe8d000c1.png)
but for lazy map,  i will suggest you try to get a supplier like this:
![image](https://user-images.githubusercontent.com/26117678/140682616-5a07f755-9a2e-4908-8d3e-321883c9424f.png)
for how to avoid double print in level 5: just try to print the supplier list once without directly calling last added supplier, although this will invoke another problem in level 3 that logger.of(x+1) will print x, but this can be avoid to modify the toString to only print the supplier instead of printing the whole supplier list
